src/System.Management.Automation/engine/PSVersionInfo.cs

```diff
@@ -571,77 +646,154 @@ public static bool TryParse(string version, out SemanticVersion result)
             }
 
-            if (parsedComponents.Length == 3 && !TryParseComponent(parsedComponents[2], "patch", ref result, out patch))
+            if (match.Groups["minor"].Success && !int.TryParse(match.Groups["minor"].Value, out minor))
             {
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            result._parsedVersion = dashIndex < 0
-                ? new SemanticVersion(major, minor, patch)
-                : new SemanticVersion(major, minor, patch, version.Substring(dashIndex + 1));
-            return true;
-        }
-
-        private static bool TryParseComponent(string component, string componentName, ref VersionResult result, out int parsedComponent)
-        {
-            if (!Int32.TryParse(component, NumberStyles.Integer, CultureInfo.InvariantCulture, out parsedComponent))
+            if (match.Groups["patch"].Success && !int.TryParse(match.Groups["patch"].Value, out patch))
             {
-                result.SetFailure(ParseFailureKind.FormatException, component);
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            if (parsedComponent < 0)
+            if (preLabel != null && !Regex.IsMatch(preLabel, LabelUnitRegEx) ||
+               (buildLabel != null && !Regex.IsMatch(buildLabel, LabelUnitRegEx)))
             {
-                result.SetFailure(ParseFailureKind.ArgumentOutOfRangeException, componentName);
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
+            result._parsedVersion = new SemanticVersion(major, minor, patch, preLabel, buildLabel);
             return true;
         }
 
         /// <summary>
-        /// ToString
+        /// Implement ToString()
         /// </summary>
         public override string ToString()
         {
-            if (Patch < 0)
+            if (versionString == null)
+            {
+                StringBuilder result = new StringBuilder();
+
+                result.Append(Major).Append(Utils.Separators.Dot).Append(Minor).Append(Utils.Separators.Dot).Append(Patch);
+
+                if (!string.IsNullOrEmpty(PreReleaseLabel))
+                {
+                    result.Append("-").Append(PreReleaseLabel);
+                }
+
+                if (!string.IsNullOrEmpty(BuildLabel))
+                {
+                    result.Append("+").Append(BuildLabel);
+                }
+
+                versionString = result.ToString();
+            }
+
+            return versionString;
+        }
+
+        /// <summary>
+        /// Implement Compare.
+        /// </summary>
+        public static int Compare(SemanticVersion versionA, SemanticVersion versionB)
+        {
+            if (versionA != null) {
+                if (versionB != null)
+                {
+                    return versionA.CompareTo(versionB);
+                }
```

Comment:
The instance method `CompareTo` handles null value in it, so this `versionB != null` check is not needed. It can just be:
```
if (versionA != null)
{
    return versionA.CompareTo(versionB);
}

if (versionB != null)
{
    return -1;
}

return 0;
```

Reply:
Fixed.